lib/layer.js

```diff
@@ -48,6 +48,31 @@ function Layer(path, options, fn) {
 }
 
 /**
+ * Handle any request for the layer.
+ *
+ * @api private
+ */
+
+Layer.prototype._handle_any = function _handle_any(/* ...args, next */) {
+  var self = this
+  var next = arguments[arguments.length - 1]
+  var handle = this.handle
+  var result
+  var hasNextBeenCalled = function () { return next._currentLayer === self }
+
+  try {
+    result = handle.apply(undefined, arguments)
```

Comment:
Basically, the `this` is defined by _where the physical `this` is located_, it will not vary depending on the call stack, which is what the issue is. In your test, the `this` for both calls is physically located in the same spot, this why it always evaluates to the same thing. My example demonstrates that the `this` is going to be physically located in _other people's code_, not within the router source code. Your use of `.apply` overrides this behavior, which is likely to result in unexpected behavior (Node.js code experienced this a lot when they tried to change the context for `setTimeout` execution).